models/shopcart.py

```diff
@@ -39,3 +37,36 @@ def all():
     def remove_all():
         """ Remove all Shopcarts from memory """
         Shopcart.__data = []
+
+    @staticmethod
+    def __validate_products(products):
+        """ Validates products or raises an error"""
+        try:
+            # Product is none so set an empty list
+            if products is None:
+                return {}
+            
+            #Not None
+            # Tuple of product, quantity 
+            if type(products) == tuple  and len(products)==2 and all(isinstance(pq,int) for pq in products): 
+                return {products[0]:products[1]}
+
+            # Just a Product id, set default quantity to 1
+            if type(products) == int: 
+                return {products:1}
+
+            if type(products) != dict : 
+                raise DataValidationError() 
+            
+            # Products is a dict of proper tuples
+            if ( all( isinstance(pid,int) for pid in products.keys() ) and
+                 all( (isinstance(q,int) and (q > 0)) for q in products.values() ) ):
+                return products
+
+            #Products not valid
+            raise DataValidationError() 
+
+        except DataValidationError as e:
```

Comment:
I think we should just raise the error with a sensible message instead of catching it and defaulting to no products? Open to conversation on this.

Reply:
I agree, but it think we can leave this for the future. Maybe we should make a new issue and pass it to the icebox. What do you guys think @RegulusXiang @radhikamattoo ??

Reply:
It also raises a message when adding products. If you raise while adding the stdoud is 
`ERROR: Products were not added to shopcart.`

Reply:
I agree with @thoughts1053 - if the products passed in are invalid, we should reject the shopcart entirely and ask them to pass in valid products. If we return no products, they technically have a valid shopcart they can perform actions on, so there may be unexpected behavior.

Reply:
I agree but this is the model not the API. So in this case it does not makes sense to ask for things again. This is not dynamic.

With this error rising we can ask the user to pass valid parameters in the API in the future.

I just tough that it does not makes sense to stop the program when adding an invalid product to an existing cart so I dont know...

Any explicit suggestions on what to do instead?

Reply:
The idea is that we raise this error on the database side with the message that the data is invalid, then the API will catch this error and construct a response for the end user. 

From a database point of view, we shouldn't have default behavior if there's invalid data - we should just reject it.  It's should be a binary decision - we either accept the valid data, or we raise an Error and reject it if any part of it is invalid. 

So instead of catching the error at all, you should just raise the errors with informative messages. Like when you raise an error if `products` is not a dictionary, you could do something like this: 

```
 if type(products) != dict : 
                 raise DataValidationError("Invalid format for products") 
```
or something similar.